### DISCUSSION TITLE
Equation with "if statements"

### DISCUSSION THREAD

**Discussion by engrmessi** [idaholab/moose#29168] [Q&A Modules: General] [answered]
Hi all, I am writing an equation (expression) that comprises "if statements" in my model as shown in the attached file. I just want to know if the equation is written correctly. if not I need correction on how best to write it in MOOSE. Thanks!
![IF_Statement](https://github.com/user-attachments/assets/5c1125ac-85e0-4382-9c53-7e8684ea4068)

**Comment by GiudGiud** [comment]
hello

no it's not. I dont know if && works so I d do this:
```
expression = 'if (t > 0, if(t<to, L0 * (1 + accslip / 0.0111), if (t<to + tf, L0 * (1 + accslip / 0.0111) * exp(-k * (t - to)), 0)), 0)'
```

**Reply by GiudGiud** [reply]
regardless of if && works, you need to keep this `if (a, b, c)` structure, even when nesting ifs: `if (a1, if(a2, b, d), c)`

**Comment by engrmessi** [comment]
Thank you for your response
But the first  part  of second inequality condition "to < t" was not accounted for. Is it Ok to ignore it? Also some of the inequality conditions has something like "less than or equal to"

**Reply by GiudGiud** [reply]
it is 
```
 if(t<to, // this part only if t < to L0 * (1 + accslip / 0.0111), // this part only if t > to if (t<to + tf,
```

**Reply by engrmessi** [reply]
![image](https://github.com/user-attachments/assets/01a9dffe-9e82-4981-97a6-36d92456975e)
This the first part. there is less than and equality sign, ie t<=to

![image](https://github.com/user-attachments/assets/663403b9-db06-4c36-957b-8fda5b564650)
There is also less than and equality sign in the second part too ie t<=to + tf

could you please take a look

**Reply by GiudGiud** [reply]
you can change < to <= in what I wrote

**Reply by engrmessi** [reply]
Yes, according to you, this is what I will have
expression = 'if (t > 0, if(t<=to, L0 * (1 + accslip / 0.001), if (t<=to + tf, L0 * (1 + accslip / 0.001) * exp(-k * (t - to)), 0)), 0)'
but my concern is that this part of inequality "to<t" in "to<t<=to + tf" is missing in the expression

**Reply by GiudGiud** [reply]
It’s not missing. 
if you don’t trust what I wrote please go to the ParsedFunction documentation then follow the link to the fparser documentation

**Reply by engrmessi** [reply]
Alright. Thanks a lot for your feedbacks, really appreciative

**Reply by engrmessi** [reply]
Good evening 
So I got the following syntax error for the expression.
expression = 'if (t > 0, if(t <= to, L0 * (1 + accslip / 0.001), if (t<=to + tf, L0 * (1 + accslip / 0.001) * exp(-k * (t - to)), 0)), 0)'. I also remove the equality sign but i got same error. Please could help see if something is missing
![error](https://github.com/user-attachments/assets/fd1e3b34-3a2b-45bc-8850-4fa0ab7e29c7)

**Reply by GiudGiud** [reply]
can you paste the rest of the input? this does not look like an error on this line, but on two lines in the input

**Reply by engrmessi** [reply]
Please is there a way I can send the input script to you privately since the project is confidential to my research lab.?

**Reply by GiudGiud** [reply]
sure. guillaume.giudicelli@inl.gov

**Reply by GiudGiud** [reply, marked as answer]
I looked at the inputs. You have in the _repassivat.i input several [/] instead of [] for the closing block. [/] is causing an error. 

I did not find any issue in diffcorr_c.i though I cannot run it because I do not have the two txt files necessary (one slip systems and one PropertyReadFile)

**Reply by engrmessi** [reply]
My code is working now
Thank you very much.

**Reply by ArashPartow** [reply]
@engrmessi  / @GiudGiud :  If the following PR  https://github.com/idaholab/moose/issues/13194 ever gets merged, you can in the future have the following expression:

```
" L :=                                                                       "
"   switch                                                                   "
"   {                                                                        "
"      case t < 0         : 0;                                               "
"      case t < to        : L0 * (1 + accslip / 0.0111);                     "
"      case t < (to + tf) : L0 * (1 + accslip / 0.0111) * exp(-k * (t - to); "
"      default            : 0;                                               "
"   };                                                                       "
"                                                                            ";
```

**Reply by engrmessi** [reply]
Good morning!
The plot and the part of script to simulate film rupture dissolution repassivation cycle FRDR for 6 cycles is attached. We are able to simulate 2, 3, 4, 5, 6 cycles by just building on the initial expression for 1 cycle. For 6 cycles, I have 6 lines of expressions. 
Now I would like to simulate for instance 100 cycles. How do I do it by not just adding 100 lines of expressions. I do not know if there is a way to use a loop to simulate for e.g. up to 1000 cycles instead of keep writing another lines of expressions. Thanks for your help

![6_Cycle FRDR](https://github.com/user-attachments/assets/5d4d15c2-8eff-4b17-bda6-dcd4022c8c1b)
![FRDR](https://github.com/user-attachments/assets/7c19ca05-2574-4d89-a62f-36c350015cfa)
![INPUT-SCRIPT](https://github.com/user-attachments/assets/58a50e1e-756d-4952-9658-6d7d7b5b6545)

**Reply by GiudGiud** [reply]
hello

well if it works like this, then you could just make a Python script that write the expression.

if that s not good enough, then I would suggest you write a custom material to do this. This is always possible in a moose application

**Reply by engrmessi** [reply]
Good morning. Thanks for suggestion. Since I am more comfortable with MATLAB, I use it to generate a MATLAB code that produces any number cyclic expression I want. My concern is how to incorporate this MATLAB code/script with my MOOSE input script. What I already tried was I created a text file (called loop.txt) in my directory and inserted the generated expression from MATLAB. Please attached files. It did not work unless i paste the equation in the MOOSE script itself (which I have been doing and i do not want to do it that way). Second thing i tried was to save the MATLAB code (called loop.m) in my directory and it did not work too.
![image](https://github.com/user-attachments/assets/166267d6-a737-4565-a71b-c17e67a6e8e2)
![image](https://github.com/user-attachments/assets/91f51f9e-632c-4edd-bd45-5db24daac12e)
![image](https://github.com/user-attachments/assets/1e944a86-4e2b-4ccc-98e8-1d8722470781)
![image](https://github.com/user-attachments/assets/a509c35a-5227-417a-8183-7bdf4172555e)

**Reply by GiudGiud** [reply]
hello

we do not support a text file for the expression parameter

how reasonable would it be to paste it in the parameter?
is it very long?
My concern is that if there are any mistakes, they would be hard to find in a very long parameter

**Reply by engrmessi** [reply]
yes it is very long expression especially when number cycle is increased. 
Maybe I will try the second option ie custom material. please how do i start with this approache?

**Reply by GiudGiud** [reply]
you ll want to create an ADMaterial
This is an example with a regular material. To make it Real -> ADReal, Material -> ADMaterial
https://mooseframework.inl.gov/getting_started/examples_and_tutorials/examples/ex08_materials.html
this workshop should help too
https://mooseframework.inl.gov/workshop/index.html#/25

**Reply by engrmessi** [reply]
Ok thank you.

**Reply by GiudGiud** [reply]
Note that parsed expression will work if correct. it will be long in the input, but it will be just-in-time compiled and will run as fast as any other ParsedMaterials

**Reply by engrmessi** [reply]
yes the parsed expression is actually working, I use Matlab to generate any desired number of cycle expression and paste it in my MOOSE script.

**Reply by GiudGiud** [reply]
so the material is just to make things cleaner?

if so just use an input file include:

!include a_file_with_just_the_material_with_the_parsed_expression.i

**Reply by engrmessi** [reply]
yes the concern is that since the expression is very long for large value of N, We want to make the main input script simple by crating a different file for the expression.
So where should this file be stored? in my job directory or in my application?

**Reply by GiudGiud** [reply]
with !include you can store it anywhere

and you dont see in the main input script it's just 1 line there!
!include ...

**Reply by GiudGiud** [reply]
https://mooseframework.inl.gov/moose/application_usage/input_syntax.html#00e4cdcb-d205-47be-9073-4b6fccb1690e

**Reply by engrmessi** [reply]
thank you for your assistance

**Reply by engrmessi** [reply]
Good afternoon
I have been running my simulation all these while with strain-controlled cyclic load, I want to switch to stress-controlled cyclic load. I am confused between "Pressure and FunctionNeumannBC" which is suitable to use. Can you recommend? the first figure is tensile cyclic stress applied at both sides and the second figure is only applied to the right side. Thank you
![stress controlled cyclic load](https://github.com/user-attachments/assets/cbebf1cf-5d72-4b41-9cd6-b2568d7e58b2)
![stress controlled load on right side](https://github.com/user-attachments/assets/66f43153-137e-47d3-8d9e-a4912f062e73)

**Reply by GiudGiud** [reply]
hello

Pressure is better for solid mechanics as it will add contributions to the Jacobian from the deformation of the mesh.

FunctionNeumannBC wont, it will just add the boundary stress applied to the residual

**Reply by engrmessi** [reply]
Thank you

**Comment by engrmessi** [comment]
Good day. I am modelling a new task where I am to apply symmetric boundary condition. Please see the figure attached. The domain is 7.2mm by 3.6mm. An initial semi elliptical pit is initiated at the center. The symmetrical boundary condition is applied at the x=3.6mm. How do I implement this symmetric BC? Thanks
![image](https://github.com/user-attachments/assets/8ffa8484-2ebe-424c-9230-b52bf3a58bf3)

**Reply by GiudGiud** [reply]
symmetry boils down to 0 flux for many equations, and 0 flux is usually the default for most equations (so no BCs would be needed)
Does it do this for your equation?

**Comment by engrmessi** [comment]
We are not using any equation, just prescribing bc along the edges using DirichletBC as applied at the left boundary as shown. My concern is applying some kind of DirichletBC at center of the domain to make it symmetrical. 
![image](https://github.com/user-attachments/assets/e1a5a2c9-af2e-4b9f-ae9a-362dc66dcaee)

**Reply by GiudGiud** [reply]
>  is applying some kind of DirichletBC at center of the domain to make it symmetrical.

a dirichletBC would not preclude symmetry but usually not necessary since the definition is 0 flux / 0 derivative (depending on the equation)

**Reply by engrmessi** [reply]
Ok thanks.
But how do I apply symmetrical BC in a domain in general?
